contracts/AstarBase.sol

```diff
@@ -0,0 +1,146 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.7;
+
+import "@openzeppelin/contracts/utils/Counters.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "./DappsStaking.sol";
+
+/// @author The Astar Network Team
+/// @title Astarbase. Mapping of Stakers ss58 <> H160
+contract AstarBase is Ownable {
+    using Counters for Counters.Counter;
+    Counters.Counter public registeredCnt;
+
+    bool public paused;
+    bool public onChainCheckEnabled;
+    address FREE_ADDRESS = address(0);
+    mapping(address => uint256) public addressMap;
+    mapping(uint256 => address) public polkadotAddrMap;
+    DappsStaking public constant DAPPS_STAKING = DappsStaking(0x0000000000000000000000000000000000005001);
+
+    constructor() {
+        paused = true;
+        onChainCheckEnabled = false;
+    }
+
+    /// @notice Register senders' address with corresponding SS58 address and store to mapping
+    /// @param polkadotAddr, SS58 address used for staking
+    function register(uint256 polkadotAddr) public {
```

Comment:
I know we discussed this but I have to comment - this is a very obvious vulnerability.
And if someone's address gets _stolen_, this will make a horrid impression of us as a company.
It could hurt us very badly.

What if we could do something like (sorry for the probably wrong syntax):
```
function register(bytes32 proof) public {
     ...
     uint256 polkadotAddr = SomePrecompile.get_address(proof);
     require(polkadotAddr != 0, "Caller doesn't own the address");
     ...
}
``` 
@hoonsubin do you know how hard would be to implement something like this?
E.g. sign some data with Polkadot.js wallet, use some precompile to verify it's legit and return SS58 address as result.

Reply:
@Dinonard that's why there is an implementation of `sudoUnRegister` where you can manually change it. If someone ends up being tricked like that, they can always open an issue and this can be fixed.
The other thing is there is no real damage from such theft. It is just the opportunity loss for missing out some airdrops or similar.

Reply:
Right now, the plan was to provide the actual proof from the client-side, meaning that users have to sign with MetaMask and Polkadot-js to register. Having said that, nothing is stopping people to call this function directly.

For your second point, I think we can implement it. We can let the contract generate a nonce.
We did something similar for our lockdrop UI during the 2nd lockdrop where the pallet will generate a nonce and the client app will try to guess it. This was used to prevent spamming transactions for a unsign transaction with chain mutation capabilities https://github.com/AstarNetwork/lockdrop-ui/blob/44eaaf562ecf1b4a2c6a76411bc221b76dd67c4c/src/helpers/plasmUtils.ts#L39.

Now, in this case, there are a couple of things the contract must be able to do.
1. Provide the nonce
2. Precompiled for signature recovery (something like https://github.com/paritytech/substrate/blob/511e8c860aa448c6899f884eebe64131a65300b8/client/cli/src/commands/verify.rs#L66 perhaps).

Then the client app can sign the static message + contact generated nonce and provide that as the `proof`, which then the contract can recover the public key and the address of the Substrate account. In this case, the `sender` will be the H160.